src/Core/Checkout/Customer/SalesChannel/LoginAsCustomerRoute.php

```diff
@@ -0,0 +1,100 @@
+use Shopware\Core\Checkout\Customer\Event\CustomerLoginEvent;
+use Shopware\Core\Checkout\Customer\Exception\CustomerNotFoundByIdException;
+use Shopware\Core\Checkout\Customer\LoginAsCustomerTokenGenerator;
+use Shopware\Core\Framework\Context;
+use Shopware\Core\Framework\DataAbstractionLayer\EntityRepository;
+use Shopware\Core\Framework\DataAbstractionLayer\Exception\InconsistentCriteriaIdsException;
+use Shopware\Core\Framework\DataAbstractionLayer\Search\Criteria;
+use Shopware\Core\Framework\Log\Package;
+use Shopware\Core\Framework\Plugin\Exception\DecorationPatternException;
+use Shopware\Core\Framework\Validation\DataBag\RequestDataBag;
+use Shopware\Core\System\SalesChannel\Context\CartRestorer;
+use Shopware\Core\System\SalesChannel\ContextTokenResponse;
+use Shopware\Core\System\SalesChannel\SalesChannelContext;
+use Symfony\Component\Routing\Annotation\Route;
+use Symfony\Contracts\EventDispatcher\EventDispatcherInterface;
+
+#[Route(defaults: ['_routeScope' => ['store-api'], '_contextTokenRequired' => false])]
+#[Package('checkout')]
+class LoginAsCustomerRoute extends AbstractLoginAsCustomerRoute
+{
+    public const CUSTOMER_ID = 'customerId';
+
+    public const SALES_CHANNEL_ID = 'salesChannelId';
+
+    public const TOKEN = 'token';
+
+    /**
+     * @internal
+     *
+     * @param EntityRepository<CustomerCollection> $customerRepository
+     */
+    public function __construct(
+        private readonly EventDispatcherInterface $eventDispatcher,
+        private readonly EntityRepository $customerRepository,
+        private readonly CartRestorer $restorer,
+        private readonly LoginAsCustomerTokenGenerator $tokenGenerator
+    ) {
+    }
+
+    public function getDecorated(): AbstractLoginAsCustomerRoute
+    {
+        throw new DecorationPatternException(self::class);
+    }
+
+    #[Route(path: '/store-api/account/login/customer', name: 'store-api.account.login-as-customer', methods: ['POST'])]
+    public function loginAsCustomer(RequestDataBag $data, SalesChannelContext $context): ContextTokenResponse
+    {
+        // TODO: find better way to handle this
+
+        if (!$data->has(self::CUSTOMER_ID)) {
+            throw CustomerException::missingCustomerId();
+        }
+
+        if (!$data->has(self::TOKEN)) {
+            throw CustomerException::missingToken();
+        }
+
+        $customerId = (string) $data->get(self::CUSTOMER_ID);
+        $token = (string) $data->get(self::TOKEN);
+
+        $this->tokenGenerator->validate($token, $context->getSalesChannelId(), $customerId);
+
+        $customer = $this->fetchCustomer($customerId, $context->getContext());
+
+        $event = new CustomerBeforeLoginEvent($context, $customer->getEmail());
+        $this->eventDispatcher->dispatch($event);
+
+        $restoredCart = $this->restorer->restore($customer->getId(), $context);
+
+        $newToken = $restoredCart->getToken();
+
+        $event = new CustomerLoginEvent($context, $customer, $newToken);
```

Comment:
@sobyte @AydinHassan
Do you think we should/could also add a new "AsCustomerLoginEvent" here so that we can differentiate/track whether the login is triggered by a standard customer login or an admin customer login event?
This would help create custom logic like an “Admin-Customer Login Log” to log when a user accesses a customer account.

To do this, it would also be nice to somehow get the user for whom the token was generated here to pass the admin user to this new event so that we can identify which user has logged into this customer account

Reply:
Maybe also the `AccountService::loginById` method should be used to login the customer: https://github.com/shopware/shopware/blob/trunk/src/Core/Checkout/Customer/SalesChannel/AccountService.php#L93

As otherwise it would be possible to login to a customer, which is for example bound to a different sales channel. But in general I think it would be useful to have different events here (not sure if it would be sufficient to only have the customer id in that event).